### PR TITLE
Implement the difference function

### LINKED ISSUE #1825: Wire up DIFFERENCE aggregate
If the query is like this:

```sql
SELECT DIFFERENCE(value)
FROM cpu
WHERE time > now() - 4h
```

It will return a series with the difference between each value.

If the query is like this:

```sql
SELECT DIFFERENCE(MIN(value))
FROM cpu
WHERE time > now() - 4h
GROUP BY time(5m)
```

It will return the difference of the min value for each 5 minute window of time.

### PR DESCRIPTION
**Pull request by jsternberg**
The difference function is implemented very similar to how derivative is
implemented. It is an aggregate function that acts over the entire
aggregate. This function will also have the same problems that
derivative has with getting values from the previous interval or point.
This will be fixed separately as part of #5943.

Fixes #1825.

### CODE REVIEW

File: cmd/influxd/run/server_test.go [hunk 1]
```diff
@@ -1948,6 +1948,219 @@ cpu value=20 1278010021000000000
 	}
 }
 
+// Ensure the server can handle various group by time difference queries.
+func TestServer_Query_SelectGroupByTimeDifference(t *testing.T) {
+	t.Parallel()
+	s := OpenServer(NewConfig())
+	defer s.Close()
+
+	if err := s.CreateDatabaseAndRetentionPolicy("db0", newRetentionPolicyInfo("rp0", 1, 1*time.Hour)); err != nil {
```
Comment (e-dard) [inline cmd/influxd/run/server_test.go]: If you have no interest in the specifics of the retention policy then you don't need this method call; `db0` and `rp0` will be created for you when `test.init` is called on each query.

File: cmd/influxd/run/server_test.go [hunk 2]
```diff
@@ -1948,6 +1948,219 @@ cpu value=20 1278010021000000000
...
+	}
+}
+
+// Ensure the server can handle various group by time difference queries.
+func TestServer_Query_SelectGroupByTimeDifferenceWithFill(t *testing.T) {
+	t.Parallel()
+	s := OpenServer(NewConfig())
+	defer s.Close()
+
+	if err := s.CreateDatabaseAndRetentionPolicy("db0", newRetentionPolicyInfo("rp0", 1, 1*time.Hour)); err != nil {
```
Comment (e-dard) [inline cmd/influxd/run/server_test.go]: Same as above.

File: cmd/influxd/run/server_test.go (hunk 1 above)
Comment (jsternberg) [inline cmd/influxd/run/server_test.go]: I just copied the setup function for `SelectGroupByTimeDerivative` which uses this for some reason.

File: influxql/ast.go [hunk 3]
```diff
@@ -1476,12 +1476,19 @@ func (s *SelectStatement) validateAggregates(tr targetRequirement) error {
...
+			case "derivative", "non_negative_derivative", "difference":
 				if err := s.validSelectWithAggregate(); err != nil {
 					return err
 				}
-				if min, max, got := 1, 2, len(expr.Args); got > max || got < min {
-					return fmt.Errorf("invalid number of arguments for %s, expected at least %d but no more than %d, got %d", expr.Name, min, max, got)
+				switch expr.Name {
+				case "derivative", "non_negative_derivative":
+					if min, max, got := 1, 2, len(expr.Args); got > max || got < min {
+						return fmt.Errorf("invalid number of arguments for %s, expected at least %d but no more than %d, got %d", expr.Name, min, max, got)
```
Comment (e-dard) [inline influxql/ast.go]: `fmt.Errorf("invalid number of arguments for %s, got %d, but wanted between %d and %d", expr.Name, got, min, max)` maybe?

File: influxql/ast.go [hunk 4]
```diff
@@ -1476,12 +1476,19 @@ func (s *SelectStatement) validateAggregates(tr targetRequirement) error {
...
+					}
+				case "difference":
+					if got := len(expr.Args); got != 1 {
+						return fmt.Errorf("invalid number of arguments for difference, expected 1, got %d", got)
```
Comment (e-dard) [inline influxql/ast.go]: switch expected and got.

File: influxql/ast.go (hunk 3 above)
Comment (jsternberg) [inline influxql/ast.go]: I don't think there's much of a difference between these and one of the error message was from a previous change. I'd prefer not to change it for consistency if possible.

File: influxql/call_iterator.go [hunk 5]
```diff
@@ -1112,3 +1116,65 @@ func NewIntegerDerivativeReduceSliceFunc(interval Interval, isNonNegative bool)
...
+			fn := NewIntegerSliceFuncReducer(IntegerDifferenceReduceSlice)
+			return fn, fn
+		}
+		return &integerReduceIntegerIterator{input: newBufIntegerIterator(input), opt: opt, create: createFn}, nil
+	default:
+		return nil, fmt.Errorf("unsupported difference iterator type: %T", input)
+	}
+}
+
+func FloatDifferenceReduceSlice(a []FloatPoint) []FloatPoint {
```
Comment (e-dard) [inline influxql/call_iterator.go]: Exported function needs a comment for golint?

File: influxql/call_iterator.go [hunk 6]
```diff
@@ -1112,3 +1116,65 @@ func NewIntegerDerivativeReduceSliceFunc(interval Interval, isNonNegative bool)
...
+		// Calculate the difference of successive points.
+		value := p.Value - prev.Value
+		prev = *p
+
+		output = append(output, FloatPoint{Time: p.Time, Value: value})
+	}
+	return output
+}
+
+func IntegerDifferenceReduceSlice(a []IntegerPoint) []IntegerPoint {
```
Comment (e-dard) [inline influxql/call_iterator.go]: Exported function needs a comment for golint?

**Comment by e-dard** [comment]
LGTM :+1

**Pull request merged**